### DISCUSSION TITLE
Quarto Dashboard Rendering without Internet Connection !!

### DISCUSSION THREAD

**Discussion by udaybmg** [quarto-dev/quarto-cli#10378] [Q&A]
### Description

Hi everyone,

I have an interesting challenge with generating Quarto reports using Python. Here’s the setup:

**Files**:

MainCode.py: Handles data collection, formatting, and generates variables for _quarto.yml.
Create_Lot_Summary.ipynb: Contains the core Python code for calculations and plots.
_quarto.yml: Configuration file for the Quarto dashboard.
Code Overview:

MainCode.py generates the _quarto.yml file with the following content:

```
data = {
    "title": "Summary for " + Lot_id,
    "output-file": Lot_id + "_Summary_" + timestamp + ".html",
    "project": {
        "output-dir": _path + "/"
    },
    "date": "today",
    "date-format": "long",
    "format": {
        "dashboard": {
            "scrolling": True,
            "link-external-newwindow": True,
            "embed-resources": False,
            "url": "./require.min.js"
        }
    },
    "execute": {
        "echo": False,
        "enabled": True
    },
    "theme": "custom.scss",
    "jupyter": "python3"
}

# Write to _quarto.yml
with open("_quarto.yml", "w") as f:
    yaml.dump(data, f)
```
After generating _quarto.yml, the following command renders the Quarto dashboard:

```
# Render Quarto to HTML
terminalcom_1 = 'quarto render Create_Lot_Summary_V1.ipynb'
subprocess.run(terminalcom_1.split(), shell=True, check=True)
```

**The Problem:**

When "embed-resources": False, the rendering creates an HTML file and a folder with supporting files.
When "embed-resources": True, the rendering creates a single HTML file with all resources embedded, but it requires an active internet connection to download files like jquery.min.js, require.min.js and may be more .

**Question:**

How can I ensure the rendering works without an internet connection, especially when "embed-resources": True?

Looking forward to your suggestions and solutions!

**Comment by cscheid** [comment]
We need an actual repository with the code to be able to help - a description like this is not enough to reproduce.

**Reply by udaybmg** [reply]
Thank you @cscheid . Your help is much appreciated.

Here is the link for [repository](https://github.com/udaybmg/Issue-with-Quarto-Dashboard-Rendering-Offline.git) with code and data file.


MainCode_V1.py: Handles data collection, formatting, and generates variables for _quarto.yml.
Dashboard_V1.ipynb: Contains the core Python code for calculations and plots.
_quarto.yml: Configuration file for the Quarto dashboard.

[A1234_Data.csv](https://github.com/udaybmg/Issue-with-Quarto-Dashboard-Rendering-Offline/blob/main/A1234_Data.csv) File has data.

file path in MainCode_V1.py. has to be adjusted accordingly to point A1234_Data.csv.

**Reply by cscheid** [reply]
If you look at the content of your ipynb files, there's no reference to jquery.min.js. That means this import is being made dynamically by one of the dependencies that you're using (I would guess it's plotly, but I'm not positive). This means there's nothing that Quarto can do about it - we can only inline dependencies that are known at time of rendering.

**Reply by udaybmg** [reply]
@cscheid Thank you for the insight! I've found that setting embed-resource='False' allows rendering to work offline. However, if the import is being made by a dependency, it should still happen even with embed-resource='False', right? I'm still curious about the underlying cause which might help to find solution. Any further ideas would be greatly appreciated!"

**Reply by cderv** [reply]
> However, if the import is being made by a dependency, it should still happen even with embed-resource='False', right? I'm still curious about the underlying cause which might help to find solution. Any further ideas would be greatly appreciated!"

You need to provide an example we can reproduce easily. Currently your code is not runnable outside of your environment. Example of error `KeyError: 'CSV_PATH'`
````
Cell In[4], line 1
----> 1 output_meta_data_csv = os.environ["CSV_PATH"]
      2 task_number = os.environ["TASK_NUMBER"]
      3 short_device_current_limit_pA = float(os.environ["short_device_current_limit_pA"])
````

Tools like `plotly` can bring lots of JS dependency. `embed_resources: true` should indeed try to make them inlined inside the HTML, but it is possible that process is missing those cells output dependencies somehow 🤔

**Reply by udaybmg** [reply]
@cderv , Thanks for looking into this! I've updated the code to correctly reference the [A1234_Data.csv](https://github.com/udaybmg/Issue-with-Quarto-Dashboard-Rendering-Offline/blob/main/A1234_Data.csv) file, which is located in the repository. I was previously unsure how to specify the repository path in the code, but I've now made the necessary changes. Please take a look at the issue once again.

**Reply by cderv** [reply]
Thanks. Though still not ok
````
ERROR: PermissionDenied: Accès refusé. (os error 5): mkdir 'C:\Users\CORK-LAB9\Downloads\Repo\Dummy_Data'
````
I can manage this one by removing in _quarto.yml, but really one advice: don't use absolute path in a project, and especially in ones you are sharing. 

Though you `.ipynb` is still calling unset env var,s ame error
````
Cell In[3], line 1
----> 1 output_meta_data_csv = os.environ["CSV_PATH"]
      2 task_number = os.environ["TASK_NUMBER"]
      3 short_device_current_limit_pA = float(os.environ["short_device_current_limit_pA"])

File <frozen os>:685, in __getitem__(self, key)

KeyError: 'CSV_PATH'
````

Please be sure to provide reproducible example.

**Reply by udaybmg** [reply]
@cderv,  Thanks a lot! I've learned a valuable lesson and will make sure to avoid using absolute paths going forward. I'll practice using relative paths or other best practices to ensure my code is more flexible and reliable.

**Reply by udaybmg** [reply]
Hi @cderv, I've updated the files in the [repository](https://github.com/udaybmg/Issue-with-Quarto-Dashboard-Rendering-Offline.git) to use relative paths. I've tested the code on a Windows PC using PyCharm IDE and it's working seamlessly. Thank you for your suggestions and patience. I apologize again for the initial inconvenience and appreciate your help in resolving the issue.

**Reply by cderv** [reply]
Cool ! 

> I've tested the code on a Windows PC using PyCharm IDE and it's working seamlessly.

Do you mean you don't have issue anymore ? Or that your reproducible example is working to show the issue ?

**Reply by udaybmg** [reply]
> Cool !
> 
> > I've tested the code on a Windows PC using PyCharm IDE and it's working seamlessly.
> 
> Do you mean you don't have issue anymore ? Or that your reproducible example is working to show the issue ?

The reproducible example is working to show the issue. In the File [MainCode_V1.py](https://github.com/udaybmg/Issue-with-Quarto-Dashboard-Rendering-Offline/blob/939ceffefb6b83e3ce014192e914e27738ea5594/MainCode_V1.py), When I set `embed-resources = False`, the rendering occurs without requiring an internet connection. However, when I set `embed-resources = True`, the rendering attempts to access the internet, and if there is no connection, it fails.

**Reply by udaybmg** [reply]
@cderv and @cscheid Checking to see if you have any solution/suggestions on the issue. Please do let me know.

**Reply by cderv** [reply]
> I've found that setting embed-resource='False' allows rendering to work offline. However, if the import is being made by a dependency, it should still happen even with embed-resource='False', right? I'm still curious about the underlying cause which might help to find solution.

When you set `embed-resources: false`, Quarto will not try to embed the resources inside the HTML file. When `embed-resources: true`, any external dependencies will be inlined in the document. If you have some python code that linked to a dependency online, it can cause your problem here: 

- `embed-resources: false`: Quarto does nothing specific. When you will open your HTML in the browser, the dependency will be fetched. You need an internet connection at viewing time ! 
- `embed-resources: true`: Quarto will see the dependency that has been brought by your python code inside the HTML file. As embedding is asked for, Quarto will do another pandoc run to embed all the resources, including online one. They will be fetched to be inlined. You need an internet connection at rendering time ! 

I hope this clarify.

**Reply by udaybmg** [reply]
Thanks @cderv .  I have tried this again and again. The results are same, When I set `embed-resources = False`, the rendering occurs without requiring an internet connection. Even to view rendered HTML file I don't need internet. Its not the case when I `embed-resources = True`, It tries to access internet even to start rendering. 

I use the script to analyze and visualize local data local, so python code doesn't need internet access.

**Reply by cderv** [reply]
> Even to view rendered HTML file I don't need internet.

Can you check you browser console and confirm that no resources are 404 and can't be loaded ? If you have no internet connection, then external resources should be erroring when being fetch by your browser. 

I will try again your example but I don't see what else could be happening. `embed-resources: true` expects an internet connection to fetch any external resources that are in the HTML file so that they are embed.

**Reply by cderv** [reply]
After some manipulation, I manage to make your example reproducible on my side. (please next time provide a ipynb that can be run without having to rely on a .py script)

First thing I can see by opening the HTML file is some `<script` loading from `https://`. So there are online resources in your HTML and when you'll try to embed then at rendering time, it requires an internet connection. 

Quarto does insert some of them (for example we insert JQuery, require.js and others like `@jupyter-widgets/html-manager` when we detect Jupyter Widgets so that they can work.)

You are also using itables which will load associated JS script from online. 

So overall you need an internet connection. Either at rendering time or at viewing time.
